src/zonetree/answer.rs

```diff
@@ -0,0 +1,217 @@
+//! Answers to zone tree queries.
+
+//------------ Answer --------------------------------------------------------
+
+use super::{SharedRr, SharedRrset, StoredDname};
+use crate::base::iana::Rcode;
+use crate::base::message_builder::AdditionalBuilder;
+use crate::base::wire::Composer;
+use crate::base::Message;
+use crate::base::MessageBuilder;
+use octseq::Octets;
+
+/// A DNS answer to a query against a [`Zone`].
+///
+/// [`Answer`] is the type returned by [`ReadableZone::query()`].
+///
+/// Callers of [`ReadableZone::query()`] will likely only ever need to use the
+/// [`Self::to_message()`] function. Alternatively, for complete control use
+/// the getter functions on [`Answer`] instead and construct a response
+/// message yourself using [`MessageBuilder`].
+///
+/// Implementers of alternate backing stores for [`Zone`]s will need to use
+/// one of the various `Answer` constructor functions when
+/// [`ReadableZone::query()`] is invoked for your zone content in order to
+/// tailor the DNS message produced by [`Self::to_message()`] based on the
+/// outcome of the query.
+///
+/// [`Zone`]: crate::zonetree::Zone
+/// [`ReadableZone::query()`]: crate::zonetree::traits::ReadableZone::query()
+#[derive(Clone)]
+pub struct Answer {
+    /// The response code of the answer.
+    rcode: Rcode,
+
+    /// The content of the answer.
+    content: AnswerContent,
+
+    /// The optional authority section to be included in the answer.
+    authority: Option<AnswerAuthority>,
+}
+
+impl Answer {
+    /// Creates an "empty" answer.
+    ///
+    /// The answer, authority and additinal sections will be empty.
+    pub fn new(rcode: Rcode) -> Self {
+        Answer {
+            rcode,
+            content: AnswerContent::NoData,
+            authority: Default::default(),
+        }
+    }
+
+    /// Creates a new message with a populated authority section.
+    ///
+    /// The answer and additional sections will be empty.
+    pub fn with_authority(rcode: Rcode, authority: AnswerAuthority) -> Self {
```

Comment:
Also, it doesn't linkify `AnswerContent` or `AnswerAuthority` in the return value position while it does linkify `Rcode` in the same position.

Reply:
It's because those types are not exported from `zonetree/mod.rs`. I suspect that the intent is that only the tree can create `Answer` and thus any setters should be private, and then those types are only used internally and not for public view.

Reply:
Actually, we need to expose those types and setter functions on `Answer` because `ReadableZone` allows a user defined backing store for `Zone` to be created and that needs to return the `Answer` type from calls to `query()`.

Reply:
Fixed.